vm/interpreter/src/lib.rs

```diff
@@ -1,130 +1,314 @@
+            act.sequence += 1;
+            Ok(())
+        })?;
 
-        // TODO: transfer gas
+        let snapshot = self.state.snapshot()?;
 
-        // TODO: increase from actor nonce
+        // scoped to deal with mutable reference borrowing
+        let (ret_data, gas_used, act_err) = {
+            let (ret_data, mut rt, act_err) = self.send(msg, msg_gas_cost as i64);
+            rt.charge_gas(rt.price_list().on_chain_return_value(ret_data.len()))
+                .map_err(|e| e.to_string())?;
+            (ret_data, rt.gas_used(), act_err)
+        };
 
-        let return_data = self.send(msg, gas_cost.to_i64().unwrap());
-        return_data
-            .map(|r| {
-                Ok(MessageReceipt {
-                    return_data: r, // TODO: what about Send?,
-                    exit_code: ExitCode::Ok,
-                    gas_used: BigUint::from(0u64), // TODO: get from runtime, runtime.gas_used()
-                })
-            })
-            .map_err(|e| {
+        if let Some(err) = act_err {
+            if err.is_fatal() {
+                return Err(format!("Fatal send actor error occurred, err: {:?}", err));
+            };
+            if err.exit_code() != ExitCode::Ok {
+                // revert all state changes since snapshot
                 if let Err(state_err) = self.state.revert_to_snapshot(&snapshot) {
-                    return state_err;
-                }
-                e.to_string()
-            })?
+                    return Err(format!("Revert state failed: {}", state_err));
+                };
+            }
+            warn!("Send actor error: from:{}, to:{}", msg.from(), msg.to());
+        }
+        let gas_used = if gas_used < 0 { 0 } else { gas_used as u64 };
+        // refund unused gas
+        let refund = (msg.gas_limit() - gas_used) * msg.gas_price();
+        self.state.mutate_actor(msg.from(), |act| {
+            act.deposit_funds(&refund);
+            Ok(())
+        })?;
+
+        let gas_reward = msg.gas_price() * BigUint::from(gas_used);
+        self.state.mutate_actor(&*REWARD_ACTOR_ADDR, |act| {
+            act.deposit_funds(&gas_reward);
+            Ok(())
+        })?;
+
+        if refund + gas_reward != gas_cost {
+            return Err("Gas handling math is wrong".to_owned());
+        }
+
+        Ok(ApplyRet::new(
+            MessageReceipt {
+                return_data: ret_data,
+                exit_code: ExitCode::Ok,
+                gas_used: BigUint::from(gas_used),
+            },
+            BigUint::zero(),
+            None,
+        ))
     }
     /// Instantiates a new Runtime, and calls internal_send to do the execution.
-    fn send(&mut self, msg: &UnsignedMessage, gas_cost: i64) -> Result<Serialized, ActorError> {
+    fn send<'m>(
+        &mut self,
+        msg: &'m UnsignedMessage,
+        gas_cost: i64,
+    ) -> (
+        Serialized,
+        DefaultRuntime<'_, 'm, '_, ST, DB, SYS>,
+        Option<ActorError>,
```

Comment:
Don't understand why this is an option? If internal_send is successful, we can return ExitCode::Ok, otherwise return actor_err, no?

Reply:
It seems weird to construct an error upon a successful call. More preference I suppose.